hazelcast/include/hazelcast/client/protocol/ClientMessage.h

```diff
@@ -795,6 +876,204 @@ class HAZELCAST_API ClientMessage
         return h;
     }
 
+    /**
+     * Reads the header of the current frame.
+     * The cursor must be at a frame's beginning.
+     */
+    frame_header_t read_frame_header()
+    {
+        frame_header_t header{};
+        auto pos = rd_ptr(SIZE_OF_FRAME_LENGTH_AND_FLAGS);
+        std::memcpy(&header.frame_len, pos, sizeof(header.frame_len));
+        pos += sizeof(header.frame_len);
+        std::memcpy(&header.flags, pos, sizeof(header.flags));
+        return header;
+    }
+
+    template<typename T>
+    typename std::enable_if<std::is_same<T, sql::sql_column_metadata>::value,
+                            T>::type
+    get()
+    {
+        // skip begin frame
+        skip_frame();
+
+        const frame_header_t header = read_frame_header();
+
+        auto type = get<int32_t>();
+
+        bool nullable = true;
+        int nullable_size = 0;
+        bool nullable_exist = false;
+        if (header.frame_len - SIZE_OF_FRAME_LENGTH_AND_FLAGS >=
+            INT32_SIZE + INT8_SIZE) {
+            nullable = get<bool>();
+            nullable_size = INT8_SIZE;
+            nullable_exist = true;
+        }
+
+        // skip bytes in initial frame
+        rd_ptr(static_cast<int32_t>(header.frame_len) -
+               SIZE_OF_FRAME_LENGTH_AND_FLAGS - INT32_SIZE - nullable_size);
+
+        std::string name = get<std::string>();
+
+        fast_forward_to_end_frame();
+
+        return codec::builtin::custom_type_factory::create_sql_column_metadata(
+          std::move(name), type, nullable_exist, nullable);
+    }
+
+    template<typename T>
+    std::vector<boost::any> to_vector_of_any(
+      std::vector<boost::optional<T>> values)
+    {
+        auto size = values.size();
+        std::vector<boost::any> vector_of_any(size);
+        for (std::size_t i = 0; i < size; ++i) {
+            auto& value = values[i];
+            if (value) {
+                vector_of_any[i] = std::move(*value);
+            }
+        }
+        return vector_of_any;
+    }
+
+    template<typename T>
+    typename std::enable_if<std::is_same<T, sql::sql_page>::value, T>::type
+    get()
+    {
+        // begin frame
+        skip_frame();
+
+        bool last = peek(SIZE_OF_FRAME_LENGTH_AND_FLAGS +
+                         1)[SIZE_OF_FRAME_LENGTH_AND_FLAGS] == 1;
+        skip_frame();
+
+        auto column_type_ids = get<std::vector<int32_t>>();
+
+        using column = std::vector<boost::any>;
+
+        auto number_of_columns = column_type_ids.size();
+        std::vector<column> columns(number_of_columns);
+        std::vector<sql::sql_column_type> column_types(number_of_columns);
+
+        for (std::size_t i = 0; i < number_of_columns; ++i) {
+            auto column_type =
+              static_cast<sql::sql_column_type>(column_type_ids[i]);
+            column_types[i] = column_type;
+
+            switch (column_type) {
```

Comment:
Would it make sense to move the switch logic to another function to make `get` shorter? Something like:
```
    std::vector<boost::any>&& convert_column(sql::sql_column_type t) {
        switch (t) {
            case sql::sql_column_type::varchar:
                return std::move(to_vector_of_any(get<std::vector<boost::optional<std::string>>>()));
            // ...
        }
    }

    template<typename T>
    typename std::enable_if<std::is_same<T, sql::sql_page>::value, T>::type
    get()
    {
         // ...
        for (std::size_t i = 0; i < number_of_columns; ++i) {
            auto column_type =
              static_cast<sql::sql_column_type>(column_type_ids[i]);
            column_types[i] = column_type;
            columns[i] = convert_column(column_type);
       }
       // ...
```

Reply:
I followed the Java code [here](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/client/impl/protocol/codec/builtin/SqlPageCodec.java#L67). But yes, it can be refactored.

I re-factored it into a method (`decode_column_values`) as suggested.